File: LeetCode/Difficulty/Medium/js/1026.maximum-difference-between-node-and-ancestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    let maxDiff = 0;

    function traverseSubtree(node,max,min){
        if(node==null || node.val==null)
            return max-min;

        max = Math.max(max, node.val);
        min = Math.min(min, node.val);

        let left = traverseSubtree(node.left,max,min);
        let right = traverseSubtree(node.right,max,min);

        return Math.max(left,right);
    }

    return traverseSubtree(root,root.val,root.val);
};

File: LeetCode/Tag/Depth-first-Search/cpp/1026.maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int V = 0;

    void updateChildrenMinMax(TreeNode* root, int& siblingMin, int& siblingMax) {
        if (!root) {
            return;
        }
        int minChild = 1e5 + 1, maxChild = -1;
        updateChildrenMinMax(root->left, minChild, maxChild);
        updateChildrenMinMax(root->right, minChild, maxChild);
        if (maxChild > -1) {
            int diff = max(abs(root->val - maxChild), abs(root->val - minChild));
            if (V < diff) {
                V = diff;
            }
        }
        siblingMin = min(siblingMin, min(root->val, minChild));
        siblingMax = max(siblingMax, max(root->val, maxChild));
    }

public:
    int maxAncestorDiff(TreeNode* root) {
        int minChild = 1e5 + 1, maxChild = -1;
        updateChildrenMinMax(root, minChild, maxChild);
        return V;
    }
};